src/irq.c

```diff
@@ -68,153 +68,167 @@ static int irq_config(const char *key, const char *value) {
-
-  if (ignorelist_match(ignorelist, irq_name) != 0)
-    return;
-
-  vl.values = &(value_t){.derive = value};
-  vl.values_len = 1;
-  sstrncpy(vl.plugin, "irq", sizeof(vl.plugin));
-  sstrncpy(vl.type, "irq", sizeof(vl.type));
-  sstrncpy(vl.type_instance, irq_name, sizeof(vl.type_instance));
+#if KERNEL_LINUX
+/* irq_strsplit is a special split function for Linux' /proc/interrupts file. It
+ * uses two or more spaces to separate fields. */
+static int irq_strsplit(char *string, char **fields, size_t fields_num) {
+  for (size_t i = 0; i < fields_num; i++) {
+    while (string[0] != 0 && isspace(string[0])) {
+      string[0] = 0;
+      string++;
+    }
 
-  plugin_dispatch_values(&vl);
-} /* void irq_submit */
+    fields[i] = string;
+    string = strstr(string, "  ");
+    if (string == NULL) {
+      return (int)i + 1;
+    }
+  }
 
-#if KERNEL_LINUX
-static int irq_read(void) {
-  FILE *fh;
-  char buffer[1024];
-  int cpu_count;
-  char *fields[256];
+  return (int)fields_num;
+}
 
+static int irq_read_data(metric_family_t *fam) {
   /*
    * Example content:
    *         CPU0       CPU1       CPU2       CPU3
    * 0:       2574          1          3          2   IO-APIC-edge      timer
    * 1:     102553     158669     218062      70587   IO-APIC-edge      i8042
    * 8:          0          0          0          1   IO-APIC-edge      rtc0
    */
-  fh = fopen("/proc/interrupts", "r");
+  FILE *fh = fopen("/proc/interrupts", "r");
   if (fh == NULL) {
     ERROR("irq plugin: fopen (/proc/interrupts): %s", STRERRNO);
     return -1;
   }
 
   /* Get CPU count from the first line */
-  if (fgets(buffer, sizeof(buffer), fh) != NULL) {
-    cpu_count = strsplit(buffer, fields, STATIC_ARRAY_SIZE(fields));
-  } else {
-    ERROR("irq plugin: unable to get CPU count from first line "
-          "of /proc/interrupts");
+  char cpu_buffer[1024] = {0};
+  if (fgets(cpu_buffer, sizeof(cpu_buffer), fh) == NULL) {
+    ERROR("irq plugin: unable to get CPU count from first line of "
+          "/proc/interrupts");
     fclose(fh);
-    return -1;
+    return EINVAL;
   }
 
+  char *cpu_names[256] = {0};
+  int cpu_count = strsplit(cpu_buffer, cpu_names, STATIC_ARRAY_SIZE(cpu_names));
+  for (int i = 0; i < cpu_count; i++) {
+    if (strncmp(cpu_names[i], "CPU", 3) == 0) {
+      cpu_names[i] += 3;
+    }
+  }
+
+  char buffer[1024] = {0};
   while (fgets(buffer, sizeof(buffer), fh) != NULL) {
-    char *irq_name;
-    size_t irq_name_len;
-    derive_t irq_value;
-    int i;
-    int fields_num;
-    int irq_values_to_parse;
-
-    fields_num = strsplit(buffer, fields, STATIC_ARRAY_SIZE(fields));
+    strstripnewline(buffer);
+
+    /* There is one column with the interrupt ID before the CPU counters and
+     * there may be up to three columns after the counters. */
+    char *fields[cpu_count + 8];
+    memset(fields, 0, sizeof(fields));
+
+    int fields_num = irq_strsplit(buffer, fields, STATIC_ARRAY_SIZE(fields));
```

Comment:
irq_strsplit() is basically same as strsplit(), except for using isspace() & manual space removal, instead of strtok().  Why it's needed?

(Gah, my comment disappeared again somewhere.)

Reply:
The difference is in the use of `strstr()`, which is used to look for two consecutive spaces. You cannot achieve this with `strtok()`, because it takes a list of characters that should act as separators, i.e. each character is considered individually rather than a substring to look for.

Reply:
Ah, I'll suggest another update to irq_strsplit() comment.